### PR TITLE
feat(plugin-api): Allow to restart a workspace using @eclipse-che/plugin

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Allow to restart a workspace from plugin API

```typescript
che.workspace.restartWorkspace(RestartWorkspaceOptions)
```

```typescript
interface RestartWorkspaceOptions: {
   prompt?: boolean;
   promptMessage? : string;
}
```

### Screenshot/screencast of this PR
restart-workspace-from-plugin

### What issues does this PR fix or reference?
N/A

### How to test this PR?
I'm providing a custom plug-in:
https://github.com/benoitf/restart-workspace-plugin

and a custom devfile with plug-in included + custom theia available:

http://che.openshift.io/f?url=https://gist.githubusercontent.com/benoitf/f211270d3b5dbfcfb8c18e0709b13f31/raw/bf4568f393d36c1ccee50fcfe9e2519c9c7bbdec/devfile-restart-workspace.yaml

Bring the command palette and search for `restart`

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

Change-Id: I4e2f17de64f972ece699016bee9c924e50cf3f0d
Signed-off-by: Florent Benoit <fbenoit@redhat.com>

### CODE REVIEW

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts [hunk 1]
```diff
@@ -39,4 +44,26 @@ export class CheWorkspaceMainImpl implements CheWorkspaceMain {
         return await this.workspaceService.updateWorkspace(workspaceId, workspace);
     }
 
+    async $restartWorkspace(machineToken: string, prompt: boolean = true): Promise<boolean> {
+        if (prompt) {
+            const confirm = new ConfirmDialog({
+                title: 'Restart Workspace',
+                msg: 'Are you sure you want to restart your workspace?',
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts:51]: I imagine a real use case when suddenly the user is asked such a question...
Doesn't it sound a little bit scary? :smile: 

Maybe add a bit more details here...?
image
WDYT?

**Review by RomanNikitenko — APPROVED** [review]
Tested using provided devfile and example of plugin, it works well - as described in the PR description.

**Review by azatsarynnyy — COMMENTED** [review]
I see a bit confusing situation in the use case when the user chooses `Restart` in Che Theia dialog but then answers `No` to the second (browser's) question.
In that case, Che Theia backend goes offline without reloading the page.
Can we somehow temporary disable the browser's confirmation immediately before reloading the page?
image

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts (hunk 1 above)
Comment (RomanNikitenko) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts:51]: maybe it makes sense to allow a plugin to provide the message for `Restart Workspace` dialog

**Comment by azatsarynnyy** [comment]
... and then restore the previous preference value after reloading, of course )

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts (hunk 1 above)
Comment (RomanNikitenko) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-workspace-main.ts:51]: so one more parameter for the method

**Review by azatsarynnyy — APPROVED** [review]

**Comment by benoitf** [comment]
[crw-ci-test]

**Pull request merged**